ouroboros-consensus-diffusion/test/consensus-test/Test/Consensus/PeerSimulator/NodeLifecycle.hs

```diff
@@ -0,0 +1,208 @@
+{-# LANGUAGE DisambiguateRecordFields #-}
+{-# LANGUAGE NamedFieldPuns           #-}
+{-# LANGUAGE ScopedTypeVariables      #-}
+
+module Test.Consensus.PeerSimulator.NodeLifecycle (
+    LiveInterval (..)
+  , LiveIntervalResult (..)
+  , LiveNode (..)
+  , LiveResources (..)
+  , NodeLifecycle (..)
+  , lifecycleStart
+  , lifecycleStop
+  , restoreNode
+  ) where
+
+import           Control.Tracer (Tracer (..), traceWith)
+import           Data.Functor (void)
+import           Data.Set (Set)
+import qualified Data.Set as Set
+import           Ouroboros.Consensus.Block
+import           Ouroboros.Consensus.Config (TopLevelConfig (..))
+import           Ouroboros.Consensus.Storage.ChainDB.API
+import qualified Ouroboros.Consensus.Storage.ChainDB.API as ChainDB
+import           Ouroboros.Consensus.Storage.ChainDB.Impl
+                     (ChainDbArgs (cdbTracer), cdbLoE)
+import qualified Ouroboros.Consensus.Storage.ChainDB.Impl as ChainDB
+import           Ouroboros.Consensus.Util.IOLike
+import           Ouroboros.Consensus.Util.ResourceRegistry
+import           Ouroboros.Network.AnchoredFragment (AnchoredFragment)
+import qualified Ouroboros.Network.AnchoredFragment as AF
+import qualified System.FS.Sim.MockFS as MockFS
+import           System.FS.Sim.MockFS (MockFS)
+import           Test.Consensus.PeerSimulator.Resources
+import           Test.Consensus.PeerSimulator.StateView
+import           Test.Consensus.PeerSimulator.Trace
+import           Test.Consensus.PointSchedule.Peers (PeerId)
+import           Test.Util.ChainDB
+import           Test.Util.Orphans.IOLike ()
+import           Test.Util.TestBlock (TestBlock, testInitExtLedger)
+
+-- | Resources used for a single live interval of the node, constructed when the
+-- node is started.
+-- When the node is shut down, 'lnCopyToImmDb' is used to persist the immutable
+-- chain.
+data LiveNode blk m = LiveNode {
+    lnChainDb          :: ChainDB m blk
+  , lnStateViewTracers :: StateViewTracers blk m
+  , lnStateTracer      :: Tracer m ()
+  , lnCopyToImmDb      :: m (WithOrigin SlotNo)
```

Comment:
What is the meaning of the returned `WithOrigin SlotNo`?

Reply:
added doc ✅